### ISSUE TITLE
question about training-regressors function

### ISSUE THREAD

**Issue by hoebat** [cnuernber/ames-house-prices#2]
with the `training-regressors` function, there are a couple of algorithms defined:

```
        base-systems [
                      ;;linear ends up blowing out the graphs.
                      ;; :daal.regression/linear
                      :daal.regression/ridge
                      :daal.regression/gradient-boosted-trees]
        base-gridsearch-systems [
                                 :smile.regression/lasso
                                 :smile.regression/support-vector
                                 :xgboost/regression]
```

why is `ml/train` used for some and `ml/gridsearch` used for others? Can just `ml/train` be used for all of them?

**Comment by cnuernber** [comment]
`ml/train` could be used for all of them but you would not get great results.  These regression models have a set of hyperparameters that you aren't seeing, especially support vector and xgboost.

For the daal library, ridge regression doesn't have any hyperparameters so there is nothing to adjust.

For daal gradient-boosted-trees, the gridsearch mechanism causes them to allocate too much memory and crash the machine.  Interestingly enough, the xgboost system which is also gradient boosted trees doesn't suffer from this drawback.

In any case, the best option is to gridsearch.  It takes a bit longer but it finds much better parameters for the models.  In order for gridsearch to work, someone has to have setup ranges for the model parameters for the gridsearch system to work with.  

Setting up the ranges looks like this:

https://github.com/techascent/tech.ml/blob/master/src/tech/libs/xgboost.clj#L170

All of those parameters are actually options in the option map used for train; this is partially explained [here](https://github.com/techascent/tech.ml#gridsearching).

So, backing up the best form of information hiding we have figured out so far with respect to training arbitrary models with arbitrary hyperparameters is gridsearching.  For some models (the daal ones), either we didn't take the time to map the model into the gridsearch system or the model is implemented such that it hard crashes with certain parameter combinations.

For the models that we have implemented auto-gridsearching, our recommended approach is certainly to gridsearch first.  Then take the top 10 models and look at the parameter ranges that made sense as that will give you a feel for the problem.  Potentially at that point gridsearching again makes sense but perhaps not.  

So, yes you can use train for all of them but you will not get as good of performance out of the system.  Gridsearching is an important tool to search through the possible parameter space for a model and that parameter space for some models is very, very large.  Understanding how each of those parameters maps to your exact problem is just not tractable across a wide range of new developers at least not as the first step.  So we provide a search mechanism to limit the amount of intuitive understand you need to get a decent first attempt.

**Comment by hoebat** [comment]
Thanks. I’ll try it out again with grid search and compare.
